File: backend/src/main/java/com/resourceplanner/service/BidService.java
package com.resourceplanner.service;

import com.resourceplanner.dto.request.BidRequest;
import com.resourceplanner.dto.response.BidResponse;
import java.util.List;
import java.util.UUID;

public interface BidService {
    List<BidResponse> getAllActive();
    BidResponse getById(UUID id);
    BidResponse create(BidRequest request);
    BidResponse update(UUID id, BidRequest request);
    BidResponse updateStatus(UUID id, String status, String reason);
    void archive(UUID id);
    void delete(UUID id);
}

File: frontend/src/app/services/bid.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Bid, BidRequest } from '../models/models';
import { environment } from '../../environments/environment';

@Injectable({ providedIn: 'root' })
export class BidService {
  private url = `${environment.apiBaseUrl}/bids`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<Bid[]> {
    return this.http.get<Bid[]>(this.url);
  }

  getById(id: string): Observable<Bid> {
    return this.http.get<Bid>(`${this.url}/${id}`);
  }

  create(request: BidRequest): Observable<Bid> {
    return this.http.post<Bid>(this.url, request);
  }

  update(id: string, request: BidRequest): Observable<Bid> {
    return this.http.put<Bid>(`${this.url}/${id}`, request);
  }

  updateStatus(id: string, status: string, reason?: string): Observable<Bid> {
    return this.http.patch<Bid>(`${this.url}/${id}/status`, null, {
      params: { status, ...(reason ? { reason } : {}) }
    });
  }

  archive(id: string): Observable<void> {
    return this.http.patch<void>(`${this.url}/${id}/archive`, {});
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.url}/${id}`);
  }
}
